### PR TITLE
[InstCombine] Replace an integer comparison of a `phi` node with multiple `ucmp`/`scmp` operands and a constant with `phi` of individual comparisons of original intrinsic's arguments

### PR DESCRIPTION
**Pull request by Poseydon42**
When we have a `phi` instruction with more than one of its incoming values being a call to `ucmp` or `scmp`, which is then compared with an integer constant, we can move the comparison through the `phi` into the incoming basic blocks because we know that a comparison of `ucmp`/`scmp` with a constant will be simplified by the next iteration of InstCombine.

This should help with one of the regressions that's blocking #107314

There's a high chance that other similar patterns can be identified, in which case they can be easily handled by the same code by moving the check for "simplifiable" instructions into a lambda.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1828]: This should also check that the CmpIntrinsic is one-use and m_Constant() should be m_APInt(), because we only fold in that case.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1814]: It looks like the logic is now duplicated between OpsToMoveUseTo and NonSimplifiedBB/InVal? Both are non-simplified cases where we move the operation to the predecessors, so we shouldn't need separate handling for the case where there is one and the case where there are multiple.

Comment (Poseydon42) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp]: Fixed, thanks

File: llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll
Comment (Poseydon42) [reply llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll:1]: The case you're proposing would still actually enable this optimization since we can still move the `icmp` into every incoming BB, and only one of the `scmp`s will not be removed due to it being used more than once (in this case the BB with `scmp` becomes what was previously `NonSimplifiedBB`). I think it's better to add a negative test where both `scmp`s are not one-use, since in that case the optimization wouldn't be trivially profitable.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1828]:
```suggestion
        match(&I, m_ICmp(m_Specific(PN), m_APInt(Ignored)))) {
```
We always place constants in RHS.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1812]:
```suggestion
  SmallVector<unsigned int> OpsToMoveUseToIncomingBB;
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

File: llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll:104]: Any tests where only 1 of the incoming values is an `{s,u}cmp` intrin?

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:1886]: Can drop braces here now.

**Review by nikic — APPROVED** [review]
LGTM

Worth noting that as implemented, the one-use check is also load-bearing for correctness, not just profitability. Otherwise we'd have to make sure we handle multi-edges correctly.

File: llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll
Comment (Poseydon42) [reply llvm/test/Transforms/InstCombine/phi-with-multiple-unsimplifiable-values.ll:134]: Added one

**Comment by Poseydon42** [comment]
> Worth noting that as implemented, the one-use check is also load-bearing for correctness, not just profitability. Otherwise we'd have to make sure we handle multi-edges correctly.

I don't think I quite understand the problem, can you give an example of a test case where removing the one-use check would cause issues?

**Comment by nikic** [comment]
> > Worth noting that as implemented, the one-use check is also load-bearing for correctness, not just profitability. Otherwise we'd have to make sure we handle multi-edges correctly.
> 
> I don't think I quite understand the problem, can you give an example of a test case where removing the one-use check would cause issues?

Something like this: https://llvm.godbolt.org/z/zEfdK4b3W The tricky part is that a predecessors occurs two times in the phi, and the value must be the same for both. So if we clone the instruction two times, we'd get a verifier error. To support this case we would need to make sure that we only clone once per duplicate predecessor.

With the one-use limitation, this is a problem we can ignore for now.

**Comment by Poseydon42** [comment]
> > > Worth noting that as implemented, the one-use check is also load-bearing for correctness, not just profitability. Otherwise we'd have to make sure we handle multi-edges correctly.
> > 
> > 
> > I don't think I quite understand the problem, can you give an example of a test case where removing the one-use check would cause issues?
> 
> Something like this: https://llvm.godbolt.org/z/zEfdK4b3W The tricky part is that a predecessors occurs two times in the phi, and the value must be the same for both. So if we clone the instruction two times, we'd get a verifier error. To support this case we would need to make sure that we only clone once per duplicate predecessor.
> 
> With the one-use limitation, this is a problem we can ignore for now.

Thank you. I'll add a comment mentioning that so there's no confusion in the future. Apart from that the patch is OK to merge, right?

**Comment by nikic** [comment]
Yes, the patch looks good to me!

**Pull request merged**
